### PR TITLE
Fix ML interfaces with `dynamic_one_shot`

### PR DESCRIPTION
**Pull request by mudit2812**
**Context:**
As name says. `qml.sample` and `qml.var` were not working when using certain interfaces due to the usage of numpy functions rather than qml.math functions.

**Description of the Change:**
* Update `dynamic_one_shot` to use `qml.math` instead of `numpy`.
* Add `qml.math.ravel` dispatch for tensorflow. This was needed for an earlier iteration for this bug fix. It is no longer being used, but is still a useful function to have.
* Updated the casting/conversion rules used when computing variance with `dynamic_one_shot`. This added a 1-10x speedup for all interfaces, with the biggest speedup for tensorflow and torch, and the smallest speedup for jax.

**Benefits:**
Better support for interfaces with `dynamic_one_shot`.

**Possible Drawbacks:**

**Related GitHub Issues:**
#5442

### CODE REVIEW

File: doc/releases/changelog-0.36.0.md [hunk 1]
```diff
@@ -551,6 +551,10 @@
 
 <h3>Bug fixes 🐛</h3>
 
+* Fixed an issue with the `dynamic_one_shot` transform where `qml.sample` and `qml.var` were not working correctly with
+  all machine learning interfaces.
+  [(#)]()
```
Comment (mudit2812) [inline doc/releases/changelog-0.36.0.md:556]:
```suggestion
  [(#5630)](https://github.com/PennyLaneAI/pennylane/pull/5630)
```

**Comment by mudit2812** [comment]
[sc-59938]

**Comment by trbromley** [comment]
@mudit2812 this is looking like a stretch for v0.36. I'm thinking to recommend this be left to the next release. Any thoughts or major worries?

**Comment by mudit2812** [comment]
> @mudit2812 this is looking like a stretch for v0.36. I'm thinking to recommend this be left to the next release. Any thoughts or major worries?

I don't think we need to push it. The fix was very small, and I'm only missing tests.

**Comment by astralcai** [comment]
Should this be rebased to master?

**Comment by trbromley** [comment]
We've decided to not target this for `v0.36`.

**Comment by mudit2812** [comment]
> Should this be rebased to master?

I'd rather leave it as it is for now as it will rebase automatically once the rc branch is merged to master. That way I can defer having to resolve merge conflicts and keep a more readable diff.

**Comment by mudit2812** [comment]
Updated fix here https://github.com/PennyLaneAI/pennylane/pull/5672

**Pull request closed without merging**